pkg/agent/agent.go

```diff
@@ -398,66 +375,48 @@ func (a *Agent) Watch(ctx context.Context, snapshotURL, diagnosticsURL string) e
 	applicationGvr, _ := schema.ParseResourceArg("applications.v1alpha1.argoproj.io")
 	applicationCallback := dc.WatchGeneric(ctx, ns, applicationGvr)
 
-	return a.watch(ctx, snapshotURL, diagnosticsURL, configAcc, coreAcc, rolloutCallback, applicationCallback)
+	return a.watch(ctx, snapshotURL, diagnosticsURL, rolloutCallback, applicationCallback)
 }
 
-type accumulator interface {
-	Changed() <-chan struct{}
-	FilteredUpdate(ctx context.Context, target interface{}, deltas *[]*kates.Delta, predicate func(*kates.Unstructured) bool) (bool, error)
-}
+func (a *Agent) waitForAPIKey(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ch := k8sapi.Subscribe(ctx, a.configWatchers.cond)
+	a.handleAPIKeyConfigChange(ctx)
 
-func (a *Agent) waitForAPIKey(ctx context.Context, configAccumulator accumulator) error {
-	isValid := func(un *kates.Unstructured) bool {
-		return true
-	}
-	configSnapshot := struct {
-		Secrets    []kates.Secret
-		ConfigMaps []kates.ConfigMap
-	}{}
 	// wait until the user installs an api key
 	for a.ambassadorAPIKey == "" {
 		select {
 		case <-ctx.Done():
-			return nil
-		case <-configAccumulator.Changed():
-			updated, err := configAccumulator.FilteredUpdate(ctx, &configSnapshot, &[]*kates.Delta{}, isValid)
-			if err != nil {
-				return err
-			}
-			if !updated {
-				continue
-			}
-			a.handleAPIKeyConfigChange(ctx, configSnapshot.Secrets, configSnapshot.ConfigMaps)
+			return
+		case <-ch:
+			a.handleAPIKeyConfigChange(ctx)
 		case <-time.After(1 * time.Minute):
 			dlog.Debugf(ctx, "Still waiting for api key")
 		}
 	}
-	return nil
 }
 
-func (a *Agent) watch(ctx context.Context, snapshotURL, diagnosticsURL string, configAccumulator accumulator, coreAccumulator accumulator, rolloutCallback <-chan *GenericCallback, applicationCallback <-chan *GenericCallback) error {
-	var err error
-	// for the watch
-	// we're not watching CRDs or anything special, so i'm pretty sure it's okay just to say all
-	// the pods are valid
-	isValid := func(un *kates.Unstructured) bool {
-		return true
-	}
+func (a *Agent) watch(ctx context.Context, snapshotURL, diagnosticsURL string, rolloutCallback <-chan *GenericCallback, applicationCallback <-chan *GenericCallback) error {
 	ambHost, err := parseAmbassadorAdminHost(snapshotURL)
 	if err != nil {
 		// if we can't parse the host out of the url we won't be able to talk to ambassador
 		// anyway
 		return err
 	}
 
+	// if emissary is not present, self-gen snapshots
+	var clusterId string
+	if a.installedBy != "emissary" {
```

Comment:
I'm not sure if this the best way of detecting if emissary is present; if the user installed the agent individually or through tp and then installs emissary this would still be false no?

Reply:
yeah. ill add the endpoint watcher as we discussed.. it wont take long